### PR TITLE
Token migration

### PR DESCRIPTION
**Pull request by lovesh**

### CODE REVIEW

File: pallets/token_migration/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+impl<T: Trait> Module<T> {
+    /// Deduct tokens from the migrator's account
+    pub fn migrate_(
+        migrator: T::AccountId,
+        recipients: BTreeMap<T::AccountId, BalanceOf<T>>,
+    ) -> dispatch::DispatchResult {
+        let allowed_migrations = Self::migrators(&migrator).unwrap();
+        let mut mig_count = recipients.len() as u16;
```
Comment (bddap) [inline pallets/token_migration/src/lib.rs:148]: Might want to guard against large recipients list (> 65535) overflowing u16.

File: pallets/token_migration/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            ensure_root(origin)?;
+            match Self::migrators(&migrator) {
+                Some(current_migrations) => {
+                    let new_migrations = current_migrations.checked_sub(decrease_migrations_by).ok_or(Error::<T>::CannotContractMigrator)?;
+                    Migrators::<T>::insert(migrator.clone(), new_migrations.clone());
+                    Self::deposit_event(RawEvent::MigratorContracted(migrator, new_migrations));
+                    Ok(())
+                },
+                None => fail!(Error::<T>::UnknownMigrator)
+            }
```
Comment (bddap) [inline pallets/token_migration/src/lib.rs:118]:
```suggestion
            let new_migrations = Self::migrators(&migrator)
                .ok_or(Error::<T>::UnknownMigrator)?
                .checked_sub(decrease_migrations_by)
                .ok_or(Error::<T>::CannotContractMigrator)?;
            Migrators::<T>::insert(&migrator, &new_migrations);
            Self::deposit_event(RawEvent::MigratorContracted(migrator, new_migrations));
            Ok(())
```

Is this sort of comment valuable to you? Feel free to mark as resolved.

File: pallets/token_migration/src/tests.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+            RawOrigin::Signed(migrator_acc).into(),
+            recips_4
+        ));
+        assert_eq!(MigrationModule::migrators(&migrator_acc).unwrap(), 0);
+        assert_eq!(
+            <TestRuntime as Trait>::Currency::free_balance(&migrator_acc).saturated_into::<u64>(),
+            0
+        );
+
+        // TODO: Check for overflow as well
```
Comment (bddap) [inline pallets/token_migration/src/tests.rs:225]: Want me to add this test?

File: pallets/token_migration/src/lib.rs (hunk 1 above)
Comment (lovesh) [reply pallets/token_migration/src/lib.rs:148]: The cast will guard by truncating > `u16::MAX` but I see how that's swallowing an error. It won't matter in practice as such a large extrinsic will exceed max length and weight allowed in a block and thus won't be included.

File: pallets/token_migration/src/lib.rs (hunk 2 above)
Comment (lovesh) [reply pallets/token_migration/src/lib.rs:118]: Yes.

File: pallets/token_migration/src/tests.rs (hunk 3 above)
Comment (lovesh) [reply pallets/token_migration/src/tests.rs:225]: I will add an issue in the board

**Pull request merged**